src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,198 @@
+        if (isEmpty()) {
+            addElementToEmptyList(node);
+        } else {
+            addElementToEndOfTheList(node);
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index: " + index + ", Size: " + size);
+        }
+        Node<T> newNode = new Node<>(value);
+        if (isEmpty()) {
+            addElementToEmptyList(newNode);
+        } else if (size == index) {
+            addElementToEndOfTheList(newNode);
+        } else if (index == 0) {
+            addElementToFirstPosition(newNode);
+        } else {
+            addElementToTheMiddle(newNode, index);
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkElementIndex(index);
+        Node<T> node = findNodeByIndex(index);
+        if (node == null) {
+            throw new IndexOutOfBoundsException("Index: " + index + " is out of bounds");
+        }
+        return node.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkElementIndex(index);
+        Node<T> node = findNodeByIndex(index);
+        T oldValue = node.value;
+        node.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkElementIndex(index);
+        Node<T> removedNode;
+        if (index == 0) {
+            removedNode = removeFirstElementFromList();
+        } else if (size == 1) {
+            removedNode = removeElementFromOneElementList();
+        } else {
+            removedNode = removeFromTheMiddleOfTheList(index);
+        }
+        size--;
+        return removedNode.value;
     }
 
     @Override
     public boolean remove(T object) {
+        int flag = 0;
+        Node<T> node;
+        for (node = first; node != null; node = node.next) {
+            if (node.value == object || (node.value != null && node.value.equals(object))) {
+                remove(flag);
+                return true;
+            }
+            flag++;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkElementIndex(int index) {
+        if (!isElementIndex(index)) {
+            throw new IndexOutOfBoundsException("Index: " + index + ", Size: " + size);
+        }
+    }
+
+    private boolean isElementIndex(int index) {
+        return index >= 0 && index < size;
+    }
+
+    private Node<T> findNodeByIndex(int index) {
+        Node<T> node;
+        if (index < (size >> 1)) {
```

Comment:
better avoid bitwise operations in order not to reduce code readability
```suggestion
        if (index < size / 2) {
```